test/indexer/test_tdr.py

```diff
@@ -218,6 +256,39 @@ def _run_sql(self, query: str) -> BigQueryRows:
     def _full_table_name(self, source: TDRSourceName, table_name: str) -> str:
         return source.bq_name + '.' + table_name
 
+    def _find_upstream_bundles(self,
```

Comment:
What's this for? It seems to be unused.

Reply:
As explained in the comment, `azul.plugins.repository.tdr.Plugin._find_upstream_bundles` performs a query that depends on features that are not implemented by TinyQuery:

```
$ git diff
diff --git a/test/indexer/test_tdr.py b/test/indexer/test_tdr.py
index 594d735f..374674e2 100644
--- a/test/indexer/test_tdr.py
+++ b/test/indexer/test_tdr.py
@@ -256,38 +256,6 @@ class TestPlugin(tdr.Plugin):
     def _full_table_name(self, source: TDRSourceName, table_name: str) -> str:
         return source.bq_name + '.' + table_name
 
-    def _find_upstream_bundles(self,
-                               source: TDRSourceRef,
-                               outputs: Entities
-                               ) -> Set[SourcedBundleFQID]:
-        # TinyQuery/legacy SQL has no support for BQ Arrays, so it's impossible
-        # to test the query in the overridden method.
-        all_links = self._run_sql(f'''
-            SELECT links_id, version, content
-            FROM {self._full_table_name(source.name, 'links')}
-        ''')
-        outputs_by_bundle: Dict[SourcedBundleFQID, Entities] = {}
-        for row in all_links:
-            bundle_fqid = SourcedBundleFQID(uuid=row['links_id'],
-                                            version=self.format_version(row['version']),
-                                            source=source)
-            bundle_outputs = {
-                EntityReference(entity_id=output['output_id'],
-                                entity_type=output['output_type'])
-                for link in json.loads(row['content'])['links']
-                if link['link_type'] == 'process_link'
-                for output in link['outputs']
-            }
-            outputs_by_bundle[bundle_fqid] = outputs & bundle_outputs
-
-        missing = outputs - set.union(*outputs_by_bundle.values())
-        require(not missing,
-                f'Dangling inputs not found in any bundle: {missing}')
-        return {
-            bundle
-            for bundle, linked_outputs in outputs_by_bundle.items()
-            if linked_outputs
-        }
 
 
 if __name__ == '__main__':
$ python3 -m unittest indexer.test_tdr.TestTDRPlugin.test_subgraph_stitching
WARNING: yacc table file version is out of date
WARNING: yacc table file version is out of date
E
======================================================================
ERROR: test_subgraph_stitching (indexer.test_tdr.TestTDRPlugin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/nadove/work/ucsc_gi/azul-sc/test/indexer/test_tdr.py", line 173, in test_subgraph_stitching
    self._test_fetch_bundle(bundle, load_tables=True)
  File "/home/nadove/.pyenv/versions/3.8.3/lib/python3.8/unittest/mock.py", line 1325, in patched
    return func(*newargs, **newkeywargs)
  File "/home/nadove/work/ucsc_gi/azul-sc/test/indexer/test_tdr.py", line 200, in _test_fetch_bundle
    emulated_bundle = plugin.fetch_bundle(test_bundle.fqid)
  File "/home/nadove/work/ucsc_gi/azul-sc/src/azul/plugins/repository/tdr/__init__.py", line 217, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
  File "/home/nadove/work/ucsc_gi/azul-sc/src/azul/plugins/repository/tdr/__init__.py", line 284, in _emulate_bundle
    entities, root_entities, links_jsons = self._stitch_bundles(bundle)
  File "/home/nadove/work/ucsc_gi/azul-sc/src/azul/plugins/repository/tdr/__init__.py", line 351, in _stitch_bundles
    upstream = self._find_upstream_bundles(source, dangling_inputs)
  File "/home/nadove/work/ucsc_gi/azul-sc/src/azul/plugins/repository/tdr/__init__.py", line 427, in _find_upstream_bundles
    for row in rows:
  File "/home/nadove/work/ucsc_gi/azul-sc/test/indexer/test_tdr.py", line 251, in _run_sql
    columns = self.tinyquery.evaluate_query(query).columns
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/tinyquery/tinyquery.py", line 257, in evaluate_query
    select_ast = compiler.compile_text(query, self.tables_by_name)
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/tinyquery/compiler.py", line 22, in compile_text
    ast = parser.parse_text(text)
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/tinyquery/parser.py", line 491, in parse_text
    return parser.parse(text, lexer=lexer.get_lexer())
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/ply/yacc.py", line 331, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/ply/yacc.py", line 1199, in parseopt_notrack
    tok = call_errorfunc(self.errorfunc, errtoken, self)
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/ply/yacc.py", line 193, in call_errorfunc
    r = errorfunc(token)
  File "/home/nadove/work/ucsc_gi/azul-sc/.venv/lib/python3.8/site-packages/tinyquery/parser.py", line 479, in p_error
    raise SyntaxError('Unexpected token: %s' % p)
  File "<string>", line None
SyntaxError: Unexpected token: LexToken(LPAREN,'(',1,168)

----------------------------------------------------------------------
Ran 1 test in 0.279s

FAILED (errors=1)
```

This overrides that method to re-implement its logic in Python rather than SQL.

Reply:
It would help to move the comment to before the method so it's more obvious that it motivates the existence of the method rather than some implementation detail of the body.

Reply:
Please also try 1) hard-coding the return value and 2) patching the method in the test, rather than overriding it.